src/backend/distributed/operations/rebalancer_placement_separation.c

```diff
@@ -0,0 +1,417 @@
+										 activeWorkerNodeList,
+										 rebalancePlacementList,
+										 shardAllowedOnNodeUDF);
+
+	return context;
+}
+
+
+/*
+ * InitRebalancerPlacementSeparationContext initializes given
+ * RebalancerPlacementSeparationContext by using given list
+ * of worker nodes and the worker node that is being drained,
+ * if specified.
+ */
+static void
+InitRebalancerPlacementSeparationContext(RebalancerPlacementSeparationContext *context,
+										 List *activeWorkerNodeList,
+										 List *rebalancePlacementList)
+{
+	HTAB *nodePlacementGroupHash = context->nodePlacementGroupHash;
+
+	List *placementListUniqueNodeGroupIds =
+		PlacementListGetUniqueNodeGroupIds(rebalancePlacementList);
+
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, activeWorkerNodeList)
+	{
+		NodeToPlacementGroupHashEntry *nodePlacementGroupHashEntry =
+			hash_search(nodePlacementGroupHash, &workerNode->groupId, HASH_ENTER,
+						NULL);
+
+		nodePlacementGroupHashEntry->shouldHaveShards =
+			workerNode->shouldHaveShards;
+		nodePlacementGroupHashEntry->hasPlacementsThatCannotBeMovedAway = false;
+		nodePlacementGroupHashEntry->assignedPlacementGroup = NULL;
+
+		/*
+		 * Lets call set of the nodes that placements in rebalancePlacementList
+		 * are stored on as D and the others as S. In other words, D is the set
+		 * of the nodes that we're allowed to move the placements "from" or
+		 * "to (*)" (* = if we're not draining it) and S is the set of the nodes
+		 * that we're only allowed to move the placements "to" but not "from".
+		 *
+		 * This means that, for a node of type S, the fact that whether the node
+		 * is used to separate a placement group or not cannot be changed in the
+		 * runtime.
+		 *
+		 * For this reason, below we find out the assigned placement groups for
+		 * nodes of type S because we want to avoid from moving the placements
+		 * (if any) from a node of type D to a node that is used to separate a
+		 * placement group within S. We also set hasPlacementsThatCannotBeMovedAway
+		 * to true for the nodes that already have some shard placements within S
+		 * because we want to avoid from moving the placements that need a separate
+		 * node (if any) from node D to node S.
+		 *
+		 * We skip below code for nodes of type D not because optimization purposes
+		 * but because it would be "incorrect" to assume that "current placement
+		 * distribution for a node of type D would be the same" after the rebalancer
+		 * plans the moves.
+		 */
+
+		if (!workerNode->shouldHaveShards)
+		{
+			/* we can't assing any shardgroup placements to the node anyway */
+			continue;
+		}
+
+		if (list_member_int(placementListUniqueNodeGroupIds, workerNode->groupId))
+		{
+			/* node is of type D */
+			continue;
+		}
+
+		/* node is of type S */
+		nodePlacementGroupHashEntry->hasPlacementsThatCannotBeMovedAway =
+			NodeGroupHasDistributedTableShardPlacements(
+				nodePlacementGroupHashEntry->nodeGroupId);
```

Comment:
> Lastly, we never set this parameter to true if we found this node in the list of placements we are rebalancing. It doesn't mean that if the node is found in that list it doesn't contain shards that are outside of this rebalance

Right, this is an outcome of the algorithm I realized last week and wishing to talk to you as well. However, I'm not sure if this would imply an important problem. Maybe we can tell users that shard separation rules may not become effective unless the rebalancer is invoked for whole cluster?

Reply:
I had a stab at getting this datastructure correctly populated: https://github.com/citusdata/citus/commit/6a64c23e8523eda0831c822604e73da57f4d7c3a